File: java/src/main/java/problems1_100/LongestSubstringWithoutRepeatingCharacters.java
package problems1_100;

import java.util.HashSet;
import java.util.Set;

enum LongestSubstringWithoutRepeatingCharacters {
    DEFAULT {
        @Override
        int lengthOfLongestSubstring(String s) {
            if (s.isEmpty()) {
                return 0;
            }
            Set<Character> characters = new HashSet<>();
            int start = 0;
            int end = 0;
            int maxLength = Integer.MIN_VALUE;
            while (end < s.length()) {
                if (characters.add(s.charAt(end))) {
                    maxLength = Math.max(maxLength, characters.size());
                    end++;
                    continue;
                }
                characters.remove(s.charAt(start));
                start++;
            }
            return maxLength;
        }
    };

    abstract int lengthOfLongestSubstring(String s);
}

File: kotlin/src/main/kotlin/problems1_100/LongestSubstringWithoutRepeatingCharacters.kt
package problems1_100

import kotlin.math.max

enum class LongestSubstringWithoutRepeatingCharacters {
    DEFAULT {
        override fun lengthOfLongestSubstring(s: String): Int {
            if (s.isEmpty()) {
                return 0
            }
            // unique characters
            val characters = hashSetOf<Char>()

            var start = 0
            var end = 0
            var maxLength = Int.MIN_VALUE
            while (end < s.length) {
                if (characters.add(s[end])) {
                    maxLength = max(maxLength, characters.size)
                    end++
                    continue
                }
                characters -= s[start]
                start++
            }
            return maxLength
        }
    },
    ;

    abstract fun lengthOfLongestSubstring(s: String): Int
}
